### PR TITLE
Custom cpu config

### PR DESCRIPTION
**Pull request by mattschlebusch**
# Reason for This PR

Provide an API endpoint that accepts JSON in an HTTP request to configure CPU features.

## Description of Changes

* API Endpoint `cpu-config`
* Sample configuration for a few CPU features
* Writes CPUID configuration
* Writes MSR entries

The above changes are tested using a firecracker process. Changes do not yet reflect a change to features for the guest. (WIP)

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/cpuid/src/template/intel/configurable_template.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+                Some(kvm_msr_entry {
+                    index: *addr,
+                    data: capabilities as u64,
+                    ..Default::default()
+                })
+            }
+        },
+        _ => None,
+    }
```
Comment (JonathanWoollett-Light) [inline src/cpuid/src/template/intel/configurable_template.rs:200]: I would suggest:
```suggestion
     if let Register::MSR { addr } = msr_pointer {
        if let Some(msr_features) = msr_enabled_features {
            let capabilities: u32 = msr_features.iter().fold(0,|acc,f eature| {
                warn!("MSR feature config - [{}] = {}", 
                    feature.config.name, feature.config.is_enabled
                );
                if feature.config.is_enabled { 
                    acc | feature.feature_mapping.bit_index 
                }
                else { 
                    acc 
                }
            });
            
            return Some(kvm_msr_entry {
                index: *addr,
                data: capabilities as u64,
                ..Default::default()
            });
        }
    }
    None
```

File: src/cpuid/src/template/intel/configurable_template.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    config: CpuConfigurationAttribute,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+enum Register {
+    EAX,
+    EBX,
+    ECX,
+    EDX,
+    MSR { addr: u32 },
```
Comment (JonathanWoollett-Light) [inline src/cpuid/src/template/intel/configurable_template.rs:233]: You could use a [newtype](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html) here.
```suggestion
enum Register {
    EAX,
    EBX,
    ECX,
    EDX,
    MSR(u32),
```

File: src/vmm/src/resources.rs [hunk 3]
```diff
@@ -241,6 +244,34 @@ impl VmResources {
...
+    pub fn configure_cpu(&mut self, cpu_config: CpuConfigurationSet) -> Result<CpuConfigError> {
+        // Validate the CPU configuration first
+        let validation_result = self.validate_cpu_configuration(&cpu_config);
+        if validation_result.is_ok() {
+            // Update the CPU configuration on the VM
+            self.vm_config.cpu_template = CUSTOM(cpu_config);
+        }
+
+        validation_result
+    }
```
Comment (JonathanWoollett-Light) [inline src/vmm/src/resources.rs:273]: Could use error propagation here like:
```suggestion
    /// Configure vCPU features.
    #[allow(unused_results)]
    pub fn configure_cpu(&mut self, cpu_config: CpuConfigurationSet) -> Result<CpuConfigError> {
        // Validate the CPU configuration first
        let validation = self.validate_cpu_configuration(&cpu_config)?; // Error propagation here.
        // Update the CPU configuration on the VM
        self.vm_config.cpu_template = CUSTOM(cpu_config);
        // Return
        validation 
    }
```

File: src/vmm/src/resources.rs [hunk 4]
```diff
@@ -286,7 +317,7 @@ impl VmResources {
         self.vm_config.mem_size_mib = mem_size_mib;
 
         // Update the CPU template
-        if let Some(cpu_template) = machine_config.cpu_template {
+        if let Some(cpu_template) = machine_config.cpu_template.clone() {
             self.vm_config.cpu_template = cpu_template;
         }
```
Comment (JonathanWoollett-Light) [inline src/vmm/src/resources.rs:322]: I find doing operations in conditions to be detrimental to readability, I would instead suggest something like:
```suggestion
        if let Some(cpu_template) = machine_config.cpu_template {
            self.vm_config.cpu_template = cpu_template.clone();
        }
```
You can also use `if let Some(cpu_template) = &machine_config.cpu_template` or `if let Some(ref cpu_template) = machine_config.cpu_template` if the aforementioned doesn't work.

**Review by JonathanWoollett-Light — CHANGES_REQUESTED** [review]

File: src/cpuid/src/cpu_config.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+}
+
+impl Default for CpuConfigurationSet {
+    fn default() -> Self {
+        CpuConfigurationSet {
+            cpu_features: Vec::new(),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply src/cpuid/src/cpu_config.rs:63]: Interesting, I got errors saying that Default did not have an implementation. I also added this quite a while ago, I will retest to confirm.

File: resources/guest_configs/cpu/cpu-config-2022-08-04.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    {
+      "name": "ibrs",
+      "is_enabled": true
+    },
+    {
+      "name": "sse4_2",
+      "is_enabled": false
+    }
+  ]
+}
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply resources/guest_configs/cpu/cpu-config-2022-08-04.json:16]: True, but the format would not be backwards compatible if we ever needed to add new fields to configuration for a given feature. (For example a specific value?) I am not confident that the `name` and `is_enabled` fields will be all that we need?

File: src/cpuid/src/template/intel/configurable_template.rs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            "sse4_2",
+            CpuFeatureArchMapping {
+                leaf: Some(crate::cpu_leaf::leaf_0x1::LEAF_NUM),
+                register: Register::EDX,
+                bit_index: crate::cpu_leaf::leaf_0x1::edx::SSE42_BITINDEX,
+                feature_type: CpuRegisterFeatureType::CPUID,
+            },
+        );
+        map
+    };
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply src/cpuid/src/template/intel/configurable_template.rs:56]: I was on the fence about using the phf crate, but I have added it for review. I do also prefer the syntax it permits.

File: src/cpuid/src/template/intel/configurable_template.rs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        match CPU_FEATURE_INDEX_MAP.get(attr.name.as_str()) {
+            None => {}
+            Some(feature_mapping) => {
+                cpu_features_list.push(CpuConfigurationInstruction {
+                    feature_mapping: feature_mapping.clone(),
+                    config: attr.clone(),
+                });
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply src/cpuid/src/template/intel/configurable_template.rs:76]: Yep a fair suggestion. I will add it to the change to test.

File: src/cpuid/src/template/intel/configurable_template.rs (hunk 2 above)
Comment (mattschlebusch) [reply src/cpuid/src/template/intel/configurable_template.rs:233]: I do find the value being named does help quite a bit, readability wise as well as syntactically.

File: src/cpuid/src/template/intel/configurable_template.rs [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+) -> () {
+    // Process CPUID features
+    // First group enabled CPU features by the leaf they are configured by
+    let leaf_to_cpuid_features_map = cpu_features
+        .into_iter()
+        .filter(|&config_instruction| {
+            config_instruction.feature_mapping.feature_type == CpuRegisterFeatureType::CPUID
+                && !config_instruction.feature_mapping.leaf.is_none()
+        })
+        .into_iter()
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply src/cpuid/src/template/intel/configurable_template.rs:94]: You are right, it builds. Will test it.

File: src/cpuid/src/template/intel/configurable_template.rs [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+    // Now build the value to be written to the MSR
+    for msr_pointer in msr_to_features_map.keys().into_iter() {
+        let msr_config_option =
+            build_single_msr_configuration(msr_pointer, msr_to_features_map.get(msr_pointer));
+
+        match msr_config_option {
+            None => (),
+            Some(kvm_msr_config) => msr_boot_entries.push(kvm_msr_config),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mattschlebusch) [reply src/cpuid/src/template/intel/configurable_template.rs:122]: Will test this out.

File: src/vmm/src/resources.rs (hunk 3 above)
Comment (mattschlebusch) [reply src/vmm/src/resources.rs:273]: I do think this is a cleaner way to go, but does require a factory of sorts for the CUSTOM type which I'm going to defer for now.

**Comment by mattschlebusch** [comment]
Substantial revisions in conjunction with new internal mechanism work ([3076](https://github.com/firecracker-microvm/firecracker/pull/3076)) are being done, this PR will be too stale to be useful at this point.

**Pull request closed without merging**